flink/src/main/java/org/apache/iceberg/flink/source/ScanContext.java

```diff
@@ -98,115 +112,203 @@ private ScanContext(boolean caseSensitive, Long snapshotId, Long startSnapshotId
+        .build();
+  }
+
+  ScanContext copyWithSnapshotId(long newSnapshotId) {
+    return ScanContext.builder()
+        .caseSensitive(caseSensitive)
+        .useSnapshotId(newSnapshotId)
+        .startSnapshotId(null)
+        .endSnapshotId(null)
+        .asOfTimestamp(null)
+        .splitSize(splitSize)
+        .splitOpenFileCost(splitOpenFileCost)
+        .nameMapping(nameMapping)
+        .streaming(isStreaming)
+        .monitorInterval(monitorInterval)
+        .build();
+  }
+
+  static Builder builder() {
+    return new Builder();
+  }
+
+  static class Builder {
+    private boolean caseSensitive;
+    private Long snapshotId;
+    private Long startSnapshotId;
+    private Long endSnapshotId;
+    private Long asOfTimestamp;
+    private Long splitSize;
+    private Integer splitLookback;
+    private Long splitOpenFileCost;
+    private boolean isStreaming;
+    private Duration monitorInterval;
+    private String nameMapping;
+    private Schema projectedSchema;
+    private List<Expression> filterExpressions;
+
+    private Builder() {
+    }
+
+    Builder caseSensitive(boolean newCaseSensitive) {
+      this.caseSensitive = newCaseSensitive;
+      return this;
+    }
+
+    Builder useSnapshotId(Long newSnapshotId) {
+      this.snapshotId = newSnapshotId;
+      return this;
+    }
+
+    Builder startSnapshotId(Long newStartSnapshotId) {
+      this.startSnapshotId = newStartSnapshotId;
+      return this;
+    }
+
+    Builder endSnapshotId(Long newEndsnapshotId) {
+      this.endSnapshotId = newEndsnapshotId;
+      return this;
+    }
+
+    Builder asOfTimestamp(Long newAsOfTimestamp) {
+      this.asOfTimestamp = newAsOfTimestamp;
+      return this;
+    }
+
+    Builder splitSize(Long newSplitSize) {
+      this.splitSize = newSplitSize;
+      return this;
+    }
+
+    Builder splitLookback(Integer newSplitLookback) {
+      this.splitLookback = newSplitLookback;
+      return this;
+    }
+
+    Builder splitOpenFileCost(Long newSplitOpenFileCost) {
+      this.splitOpenFileCost = newSplitOpenFileCost;
+      return this;
+    }
+
+    Builder streaming(boolean streaming) {
+      this.isStreaming = streaming;
+      return this;
+    }
+
+    Builder monitorInterval(Duration newMonitorInterval) {
+      this.monitorInterval = newMonitorInterval;
+      return this;
+    }
+
+    Builder nameMapping(String newNameMapping) {
+      this.nameMapping = newNameMapping;
+      return this;
+    }
+
+    Builder projectedSchema(Schema newProjectedSchema) {
+      this.projectedSchema = newProjectedSchema;
+      return this;
+    }
+
+    Builder filterExpression(List<Expression> newFilterExpressions) {
+      this.filterExpressions = newFilterExpressions;
+      return this;
+    }
+
+    Builder fromProperties(Map<String, String> properties) {
+      Configuration config = new Configuration();
+      properties.forEach(config::setString);
+
+      return new Builder()
```

Comment:
I think returning a different builder breaks the builder pattern. Why not call these methods on `this` instead?

Reply:
Yeah, you are right !  If use a new builder, then it should be copying the builder , maybe named it as `mergeProperties`, here we should use `this` .